rust/numaflow-core/src/config/components.rs

```diff
@@ -619,6 +619,192 @@ pub(crate) mod metrics {
+        Session(SessionWindowConfig),
+        Accumulator(AccumulatorWindowConfig),
+    }
+
+    #[derive(Debug, Clone, PartialEq)]
+    pub(crate) struct FixedWindowConfig {
+        pub(crate) length: Duration,
+    }
+
+    impl From<Box<FixedWindow>> for FixedWindowConfig {
+        fn from(value: Box<FixedWindow>) -> Self {
+            Self {
+                length: value.length.map(Duration::from).unwrap_or_default(),
+            }
+        }
+    }
+
+    #[derive(Debug, Clone, PartialEq)]
+    pub(crate) struct SlidingWindowConfig {
+        pub(crate) length: Duration,
+        pub(crate) slide: Duration,
+    }
+
+    impl From<Box<SlidingWindow>> for SlidingWindowConfig {
+        fn from(value: Box<SlidingWindow>) -> Self {
+            Self {
+                length: value.length.map(Duration::from).unwrap_or_default(),
+                slide: value.slide.map(Duration::from).unwrap_or_default(),
+            }
+        }
+    }
+
+    #[derive(Debug, Clone, PartialEq)]
+    pub(crate) struct SessionWindowConfig {
+        pub(crate) timeout: Duration,
+    }
+
+    impl From<Box<SessionWindow>> for SessionWindowConfig {
+        fn from(value: Box<SessionWindow>) -> Self {
+            Self {
+                timeout: value.timeout.map(Duration::from).unwrap_or_default(),
+            }
+        }
+    }
+
+    #[derive(Debug, Clone, PartialEq)]
+    pub(crate) struct AccumulatorWindowConfig {
+        pub(crate) timeout: Duration,
+    }
+
+    impl From<Box<AccumulatorWindow>> for AccumulatorWindowConfig {
+        fn from(value: Box<AccumulatorWindow>) -> Self {
+            Self {
+                timeout: value.timeout.map(Duration::from).unwrap_or_default(),
+            }
+        }
+    }
+
+    impl TryFrom<&Box<GroupBy>> for WindowConfig {
+        type Error = Error;
+        fn try_from(group_by: &Box<GroupBy>) -> Result<Self> {
+            let window = group_by.window.as_ref();
+
+            let window_type = if let Some(fixed) = &window.fixed {
+                WindowType::Fixed(fixed.clone().into())
+            } else if let Some(sliding) = &window.sliding {
+                WindowType::Sliding(sliding.clone().into())
+            } else if let Some(session) = &window.session {
+                WindowType::Session(session.clone().into())
+            } else if let Some(accumulator) = &window.accumulator {
+                WindowType::Accumulator(accumulator.clone().into())
+            } else {
+                return Err(Error::Config("No window type specified".to_string()));
+            };
+
+            Ok(WindowConfig {
+                window_type,
+                allowed_lateness: group_by
+                    .allowed_lateness
+                    .map_or(Duration::from_secs(0), Duration::from),
+                is_keyed: group_by.keyed.unwrap_or(false),
+            })
+        }
+    }
+
+    #[derive(Debug, Clone, PartialEq)]
+    pub(crate) struct StorageConfig {
```

Comment:
this is not called WAL because this can be generic for any storage, right?

Reply:
yes